### ISSUE TITLE
Why ``furl.set`` call ``self.query.load`` not ``self.query.set``?

### ISSUE THREAD

**Issue by Brightcells** [gruns/furl#95]
```python
In [1]: furl('http://www.google.com/?one=1&two=2').set({'one': 'one'})
Out[1]: furl('http://www.google.com/?one=one')

In [2]: furl('http://www.google.com/?one=1&two=2').remove('one').add({'one': 'one'})
Out[2]: furl('http://www.google.com/?two=2&one=one')
```
How to realize ``In [2]`` more simply

**Comment by gruns** [comment]
> Why ``furl.set`` call ``self.query.load`` not ``self.query.set``?

Because the scope of `furl.set()`'s parameters and `furl.query.set()`'s
parameters are different: `furl.set()` operates at the scope of the entire URL
while `furl.query.set()` operates at the scope of the query.

`furl.set(query=...)`'s `query` parameter represents the entire query of a URL,
hence it calls `Query.load()` instead of `Query.set()`. In English,
`furl.set(query=...)` translates to

> Set the query of the URL to the provided value.

For example, one can also provide an entire query string to `furl.set()`, too:

```python
>>> from furl import furl
>>> f = furl('http://www.google.com/')
>>> f.set(query='a=1&b=2')
furl('http://www.google.com/?a=1&b=2')
>>> f.args
omdict1D([('a', '1'), ('b', '2')])
```

On the other hand, `Query.set(mapping=...)`'s `mapping` paramter is a key:value
mapping of key:value pairs to adopt. In English, this translates to

> Set the keys and values of the Query to those in the provided mapping.

Does that answer your question?

**Comment by Brightcells** [comment]
@gruns Thanks a lot, your answer is pretty good to me.

```python
In [2]: furl('http://www.google.com/?one=1&two=2').remove('one').add({'one': 'one'})
Out[2]: furl('http://www.google.com/?two=2&one=one')
```

What about adding a func for ``furl`` to realize ``remove('one').add({'one': 'one'})``
When query amount is multi, this will be very useful.

Now I realize this as below, should to consider ``None``. NOT PYTHONIC.
```python
In [4]: query = {'one': 'one', 'two': 'two'}

In [5]: furl('http://www.google.com/?one=1&two=2').remove(query.keys()).add(query)
Out[5]: furl('http://www.google.com/?two=two&one=one')

In [6]: query = {}

In [7]: furl('http://www.google.com/?one=1&two=2').remove(query.keys()).add(query)
Out[7]: furl('http://www.google.com/?one=1&two=2')

In [8]: query = None

In [9]: furl('http://www.google.com/?one=1&two=2').remove(query.keys()).add(query)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-9-f731add82b89> in <module>()
----> 1 furl('http://www.google.com/?one=1&two=2').remove(query.keys()).add(query)

AttributeError: 'NoneType' object has no attribute 'keys'
```

**Comment by gruns** [comment]
> What about adding a func for furl to realize remove('one').add({'one': 'one'})
> When query amount is multi, this will be very useful.

There's another, straightforward way to accomplish the same task without
enumerating the keys twice: remove the entire query with
`furl.remove(query=True)`.

For example:

```python
>>> from furl import furl
>>> f = furl('http://www.google.com/?one=1&two=2')
>>> f.remove(query=True).set({'new': 'arg'}).url
'http://www.google.com/?new=arg'
```

Does `furl.remove(query=True).set({...})` suffice for your needs?

**Comment by Brightcells** [comment]
Remove entire query does not suffice for my needs.

In general, I just need to replace the query when the key exists, to add the query when the key not exists.

Because if I directly add the query, sometimes the query key will be multi.

Django for example:

* ``request.GET.get('key', '')`` will not get the result
* ``request.GET.get('key[]', '')`` will get the result

Then this will raise some bugs.

**Comment by gruns** [comment]
> Remove entire query does not suffice for my needs.
> 
> In general, I just need to replace the query when the key exists, to
> add the query when the key not exists.
>
> Because if I directly add the query, sometimes the query key will be
> multi.

`furl.query.set()` performs exactly the actions you describe. If the
query key already exists, `furl.query.set()` replaces it (even
multivalue keys, like `?one=a&one=b`).

```python
>>> from furl import furl
>>> f = furl('http://www.google.com/?a=a&b=1&b=2')
>>> f.query.set({'b':'REPLACED'})
>>> f.url
'http://www.google.com/?a=a&b=REPLACED'
```

And if the query key doesn't exist, `furl.query.set()` adds it.

```python
>>> from furl import furl
>>> f = furl('http://www.google.com/?a=a')
>>> f.query.set({'b':'ADDED'})
>>> f.url
'http://www.google.com/?a=a&b=ADDED'
```

So, to back up a bit and make sure we're on the same page:
`furl.query.set()` does exactly what you want -- you'd just like a
similar method on the `furl` object, too, not just the `Query` object,
right? Something like

```python
>>> from furl import furl
>>> f = furl('http://www.google.com/?a=a&b=b')
>>> f.set_query({'b':'REPLACED'})
>>> f.url
'http://www.google.com/?a=a&b=REPLACED'
```

Do I understand you correctly?

**Comment by Brightcells** [comment]
Yeah, you got it.
Use ``Query`` object does exactly what I want except cannot on a chain.
What about adding a similar method on the furl object

**Comment by gruns** [comment]
Wonderful. Now I understand exactly what you want.

That said, I do not think method chaining between `furl` objects and their URL
components (`Path`, `Query`, and `Fragment`) is common enough nor profound
enough to warrant additional API complexity. Furl's existing API, without
inter-URL-component method chaining, is clear enough and Pythonic enough.

```python
>>> from furl import furl
>>> f = furl('http://www.google.com/?a=a&b=b')
>>> f.query.set({'b':'REPLACED})
>>> # Do something with the original furl object <f>...
```

That, and the addition of new parameters

```python
>>> f.set(query_set=...)
```

methods

```python
>>> f.query_set(...)
```

and/or attributes

```python
>>> f.query.set(...).parent.url
```

isn't worth the additional API foot print.

If many others want similar method chaining tree, I will reconsider this
issue. If you are one such other person, please feel free to re-open this issue.

In the mean time, if you absolutely, positively must method chain Query
modifications into `furl.set()`, you can provide `query=str(f.query.set(...))`
to `furl.set()`. Like

```python
>>> f = furl('http://www.google.com/?a=a&b=b')
>>> f.set(query=str(f.query.set({'b':'REPLACED'}))).url
'http://www.google.com/?a=a&b=REPLACED'
```

This is, of course, ugly and not recommended.

In conclusion, I hope my explanations and thoughts help. Thank you for bringing
this issue to my attention answering my questions, Kimi. If you have anything to
add to this issue or otherwise, don't hesitate to let me know.
